File: tensorflow/contrib/metrics/python/ops/set_ops.py
# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Python layer for set_ops."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from tensorflow.contrib.framework.python.framework import tensor_util

from tensorflow.python.framework import dtypes
from tensorflow.python.framework import load_library
from tensorflow.python.framework import ops
from tensorflow.python.framework import tensor_shape
from tensorflow.python.platform import resource_loader


_set_ops = load_library.load_op_library(
    resource_loader.get_path_to_datafile("_set_ops.so"))
assert _set_ops, "Could not load _set_ops.so."


_VALID_DTYPES = set([
    dtypes.int8, dtypes.int16, dtypes.int32, dtypes.int64,
    dtypes.uint8, dtypes.uint16, dtypes.string])


@ops.RegisterShape("SetSize")
def _size_shape(unused_op):
  """Shape function for SetSize op."""
  return [tensor_shape.unknown_shape()]


def set_size(a, validate_indices=True):
  """Compute number of unique elements along last dimension of `a`.

  Args:
    a: `SparseTensor`, with indices sorted in row-major order.
    validate_indices: Whether to validate the order and range of sparse indices
       in `a`.

  Returns:
    For `a` ranked `n`, this is a `Tensor` with rank `n-1`, and the same 1st
    `n-1` dimensions as `a`. Each value is the number of unique elements in
    the corresponding `[0...n-1]` dimension of `a`.

  Raises:
    TypeError: If `a` is an invalid types.
  """
  a = tensor_util.convert_to_tensor_or_sparse_tensor(a, name="a")
  if not isinstance(a, ops.SparseTensor):
    raise TypeError("Expected `SparseTensor`, got %s." % a)
  if a.values.dtype.base_dtype not in _VALID_DTYPES:
    raise TypeError("Invalid dtype %s." % a.values.dtype)
  # pylint: disable=protected-access
  return _set_ops.set_size(a.indices, a.values, a.shape, validate_indices)

ops.NoGradient("SetSize")


@ops.RegisterShape("DenseToDenseSetOperation")
def _dense_to_dense_shape(op):
  """Shapes for `SparseTensor` result given 2 dense inputs.

  Args:
    op: Operation with 2 dense `Tensor` inputs.

  Returns:
    Tuple of three shapes corresponding to the indices, values, and shape
    `Tensor` components of the result `SparseTensor`.

  Raises:
    ValueError: if either input `Tensor` has rank < 2, or ranks do not match, or
    first n-1 dims of input shapes are not compatible.
  """
  # The following should stay in sync with `ComputeDenseToDense` shape
  # assertions in kernels/set_kernels.cc.
  input0_shape = op.inputs[0].get_shape()
  input0_rank = input0_shape.ndims
  if (input0_rank is not None) and (input0_rank < 2):
    raise ValueError("Input 0, expected rank >= 2, got shape %s." %
                     input0_shape)
  # Dimension n contains the set values to be compared, so ranks and the first
  # n-1 dimensions of inputs and output must match.
  input1_shape = op.inputs[1].get_shape()
  input1_rank = input1_shape.ndims
  if (input0_rank is not None) and (input1_rank is not None) and (
      input0_rank != input1_rank):
    raise ValueError(
        "Ranks do not match: input 0 with shape %s, input 1 with shape %s." %
        (input0_shape, input1_shape))
  output_rank = input1_rank if input0_rank is None else input0_rank
  output_dim0 = input1_shape[1] if input0_shape[0] is None else input0_shape[0]
  input0_dims = input0_shape.dims
  if input0_dims is None:
    group0_shape = tensor_shape.unknown_shape()
  else:
    group0_shape = tensor_shape.TensorShape(input0_dims[:-1])
  input1_dims = input1_shape.dims
  if input1_dims is None:
    group1_shape = tensor_shape.unknown_shape()
  else:
    group1_shape = tensor_shape.TensorShape(input1_dims[:-1])
  group0_shape.assert_is_compatible_with(group1_shape)

  indices_shape = tensor_shape.TensorShape((output_dim0, output_rank))
  values_shape = tensor_shape.unknown_shape(1)
  shape_shape = tensor_shape.TensorShape((output_rank,))
  return (indices_shape, values_shape, shape_shape)


@ops.RegisterShape("DenseToSparseSetOperation")
def _dense_to_sparse_shape(op):
  """Shapes for `SparseTensor` result given 1 dense input and 1 sparse input.

  Args:
    op: Operation with 1 dense `Tensor` and 1 `SparseTensor` input.

  Returns:
    Tuple of three shapes corresponding to the indices, values, and shape
    `Tensor` components of the result `SparseTensor`.

  Raises:
    ValueError: if either input `Tensor` has rank < 2.
  """
  # The following should stay in sync with `ComputeDenseToSparse` shape
  # assertions in kernels/set_kernels.cc.
  input_shape = op.inputs[0].get_shape()
  input_rank = input_shape.ndims
  if (input_rank is not None) and (input_rank < 2):
    raise ValueError("Expected rank >= 2, got %s." % input_shape)
  # Assert valid dimensions for the 3 `Tensor` components of `SparseTensor`.
  ops.SparseTensor(op.inputs[1], op.inputs[2], op.inputs[3])

  indices_shape = tensor_shape.TensorShape((input_shape[0], input_rank))
  values_shape = tensor_shape.unknown_shape(1)
  shape_shape = tensor_shape.TensorShape((input_rank,))
  return (indices_shape, values_shape, shape_shape)


@ops.RegisterShape("SparseToSparseSetOperation")
def _sparse_to_sparse_shape(op):
  """Shapes for `SparseTensor` result given 2 sparse inputs.

  Args:
    op: Operation with 2 `SparseTensor` inputs.

  Returns:
    Tuple of three shapes corresponding to the indices, values, and shape
    `Tensor` components of the result `SparseTensor`.
  """
  # The following should stay in sync with `ComputeSparseToSparse` shape
  # assertions in kernels/set_kernels.cc.
  # Assert valid dimensions for the 3 `Tensor` components of `SparseTensor`.
  ops.SparseTensor(op.inputs[0], op.inputs[1], op.inputs[2])
  ops.SparseTensor(op.inputs[3], op.inputs[4], op.inputs[5])

  indices_shape = tensor_shape.unknown_shape(2)
  values_shape = tensor_shape.unknown_shape(1)
  shape_shape = tensor_shape.unknown_shape(1)
  return (indices_shape, values_shape, shape_shape)


ops.NoGradient("DenseToDenseSetOperation")
ops.NoGradient("DenseToSparseSetOperation")
ops.NoGradient("SparseToSparseSetOperation")


def _set_operation(a, b, set_operation, validate_indices=True):
  """Compute set operation of elements in last dimension of `a` and `b`.

  All but the last dimension of `a` and `b` must match.

  Args:
    a: `Tensor` or `SparseTensor` of the same type as `b`. If sparse, indices
        must be sorted in row-major order.
    b: `Tensor` or `SparseTensor` of the same type as `a`. Must be
        `SparseTensor` if `a` is `SparseTensor`. If sparse, indices must be
        sorted in row-major order.
    set_operation: String indicating set operaiton. See
        SetOperationOp::SetOperationFromContext for valid values.
    validate_indices: Whether to validate the order and range of sparse indices
       in `a` and `b`.

  Returns:
    A `SparseTensor` with the same rank as `a` and `b`, and all but the last
    dimension the same. Elements along the last dimension contain the results
    of the set operation.

  Raises:
    TypeError: If inputs are invalid types.
    ValueError: If `a` is sparse and `b` is dense.
  """
  a = tensor_util.convert_to_tensor_or_sparse_tensor(a, name="a")
  if a.dtype.base_dtype not in _VALID_DTYPES:
    raise TypeError("'a' invalid dtype %s." % a.dtype)
  b = tensor_util.convert_to_tensor_or_sparse_tensor(b, name="b")
  if b.dtype.base_dtype != a.dtype.base_dtype:
    raise TypeError("Types don't match, %s vs %s." % (a.dtype, b.dtype))
  # pylint: disable=protected-access
  if isinstance(a, ops.SparseTensor):
    if isinstance(b, ops.SparseTensor):
      indices, values, shape = _set_ops.sparse_to_sparse_set_operation(
          a.indices, a.values, a.shape, b.indices, b.values, b.shape,
          set_operation, validate_indices)
    else:
      raise ValueError("Sparse,Dense is not supported, but Dense,Sparse is. "
                       "Please flip the order of your inputs.")
  elif isinstance(b, ops.SparseTensor):
    indices, values, shape = _set_ops.dense_to_sparse_set_operation(
        a, b.indices, b.values, b.shape, set_operation, validate_indices)
  else:
    indices, values, shape = _set_ops.dense_to_dense_set_operation(
        a, b, set_operation, validate_indices)
  # pylint: enable=protected-access
  return ops.SparseTensor(indices, values, shape)


def set_intersection(a, b, validate_indices=True):
  """Compute set intersection of elements in last dimension of `a` and `b`.

  All but the last dimension of `a` and `b` must match.

  Args:
    a: `Tensor` or `SparseTensor` of the same type as `b`. If sparse, indices
        must be sorted in row-major order.
    b: `Tensor` or `SparseTensor` of the same type as `a`. Must be
        `SparseTensor` if `a` is `SparseTensor`. If sparse, indices must be
        sorted in row-major order.
    validate_indices: Whether to validate the order and range of sparse indices
       in `a` and `b`.

  Returns:
    A `SparseTensor` with the same rank as `a` and `b`, and all but the last
    dimension the same. Elements along the last dimension contain the
    intersections.
  """
  return _set_operation(a, b, "intersection", validate_indices)


def set_difference(a, b, aminusb=True, validate_indices=True):
  """Compute set difference of elements in last dimension of `a` and `b`.

  All but the last dimension of `a` and `b` must match.

  Args:
    a: `Tensor` or `SparseTensor` of the same type as `b`. If sparse, indices
        must be sorted in row-major order.
    b: `Tensor` or `SparseTensor` of the same type as `a`. Must be
        `SparseTensor` if `a` is `SparseTensor`. If sparse, indices must be
        sorted in row-major order.
    aminusb: Whether to subtract `b` from `a`, vs vice versa.
    validate_indices: Whether to validate the order and range of sparse indices
       in `a` and `b`.

  Returns:
    A `SparseTensor` with the same rank as `a` and `b`, and all but the last
    dimension the same. Elements along the last dimension contain the
    differences.
  """
  return _set_operation(a, b, "a-b" if aminusb else "b-a", validate_indices)


def set_union(a, b, validate_indices=True):
  """Compute set union of elements in last dimension of `a` and `b`.

  All but the last dimension of `a` and `b` must match.

  Args:
    a: `Tensor` or `SparseTensor` of the same type as `b`. If sparse, indices
        must be sorted in row-major order.
    b: `Tensor` or `SparseTensor` of the same type as `a`. Must be
        `SparseTensor` if `a` is `SparseTensor`. If sparse, indices must be
        sorted in row-major order.
    validate_indices: Whether to validate the order and range of sparse indices
       in `a` and `b`.

  Returns:
    A `SparseTensor` with the same rank as `a` and `b`, and all but the last
    dimension the same. Elements along the last dimension contain the
    unions.
  """
  return _set_operation(a, b, "union", validate_indices)

File: tensorflow/contrib/metrics/ops/set_ops.cc
/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"

namespace tensorflow {

using shape_inference::DimensionHandle;
using shape_inference::InferenceContext;
using shape_inference::ShapeHandle;

REGISTER_OP("SetSize")
    .Input("set_indices: int64")
    .Input("set_values: T")
    .Input("set_shape: int64")
    .Attr("validate_indices: bool = true")
    .Attr("T: {int8, int16, int32, int64, uint8, uint16, string}")
    .Output("size: int32")
    .SetShapeFn(shape_inference::UnknownShape)
    .Doc(R"doc(
Number of unique elements along last dimension of input `set`.

Input `set` is a `SparseTensor` represented by `set_indices`, `set_values`,
and `set_shape`. The last dimension contains values in a set, duplicates are
allowed but ignored.

If `validate_indices` is `True`, this op validates the order and range of `set`
indices.

set_indices: 2D `Tensor`, indices of a `SparseTensor`.
set_values: 1D `Tensor`, values of a `SparseTensor`.
set_shape: 1D `Tensor`, shape of a `SparseTensor`.
size: For `set` ranked `n`, this is a `Tensor` with rank `n-1`, and the same 1st
    `n-1` dimensions as `set`. Each value is the number of unique elements in
    the corresponding `[0...n-1]` dimension of `set`.
)doc");

REGISTER_OP("DenseToDenseSetOperation")
    .Input("set1: T")
    .Input("set2: T")
    .Attr("set_operation: string")
    .Attr("validate_indices: bool = true")
    .Attr("T: {int8, int16, int32, int64, uint8, uint16, string}")
    .Output("result_indices: int64")
    .Output("result_values: T")
    .Output("result_shape: int64")
    .SetShapeFn([](InferenceContext* c) {
      if (c->num_inputs() != 2) {
        return errors::InvalidArgument("len(inputs) != 2.");
      }
      // The following should stay in sync with `ComputeDenseToDense` shape
      // assertions in kernels/set_kernels.cc.
      // Dimension n contains the set values to be compared, so ranks and the
      // first n-1 dimensions of inputs and output must match.
      DimensionHandle output_rank;
      ShapeHandle input0_shape = c->input(0);
      if (c->RankKnown(input0_shape)) {
        const int32 input0_rank = c->Rank(input0_shape);
        if (input0_rank < 2) {
          return errors::InvalidArgument("Input 0, expected rank >= 2, got ",
                                         input0_rank, ".");
        }
        ShapeHandle input1_shape = c->input(1);
        if (c->RankKnown(input1_shape)) {
          const int32 rank = c->Rank(input1_shape);
          if (input0_rank != rank) {
            return errors::InvalidArgument("Ranks do not match: input 0 ",
                                           input0_rank, ", input 1 ", rank,
                                           ".");
          }
          ShapeHandle group0_shape;
          TF_RETURN_IF_ERROR(
              c->Subshape(input0_shape, 0, rank - 1, &group0_shape));
          ShapeHandle group1_shape;
          TF_RETURN_IF_ERROR(
              c->Subshape(input1_shape, 0, rank - 1, &group1_shape));
          ShapeHandle unused_shape;
          TF_RETURN_IF_ERROR(
              c->Merge(group0_shape, group1_shape, &unused_shape));
        }
        output_rank = c->MakeDim(input0_rank);
      } else {
        ShapeHandle input1_shape = c->input(1);
        if (c->RankKnown(input1_shape)) {
          const int32 input1_rank = c->Rank(input1_shape);
          if (input1_rank < 2) {
            return errors::InvalidArgument("Input 0, expected rank >= 2, got ",
                                           input1_rank, ".");
          }
          output_rank = c->MakeDim(input1_rank);
        } else {
          output_rank = c->UnknownDim();
        }
      }
      DimensionHandle output_num_elements = c->Dim(input0_shape, 0);
      if (!c->ValueKnown(output_num_elements)) {
        ShapeHandle input1_shape = c->input(1);
        output_num_elements = c->Dim(input1_shape, 0);
        if (!c->ValueKnown(output_num_elements)) {
          output_num_elements = c->UnknownDim();
        }
      }

      c->set_output(0, c->Matrix(output_num_elements, output_rank));
      c->set_output(1, c->Vector(output_num_elements));
      c->set_output(2, c->Vector(output_rank));
      return Status::OK();
    })
    .Doc(R"doc(
Applies set operation along last dimension of 2 `Tensor` inputs.

See SetOperationOp::SetOperationFromContext for values of `set_operation`.

Output `result` is a `SparseTensor` represented by `result_indices`,
`result_values`, and `result_shape`. For `set1` and `set2` ranked `n`, this
has rank `n` and the same 1st `n-1` dimensions as `set1` and `set2`. The `nth`
dimension contains the result of `set_operation` applied to the corresponding
`[0...n-1]` dimension of `set`.

set1: `Tensor` with rank `n`. 1st `n-1` dimensions must be the same as `set2`.
    Dimension `n` contains values in a set, duplicates are allowed but ignored.
set2: `Tensor` with rank `n`. 1st `n-1` dimensions must be the same as `set1`.
    Dimension `n` contains values in a set, duplicates are allowed but ignored.
result_indices: 2D indices of a `SparseTensor`.
result_values: 1D values of a `SparseTensor`.
result_shape: 1D `Tensor` shape of a `SparseTensor`. `result_shape[0...n-1]` is
    the same as the 1st `n-1` dimensions of `set1` and `set2`, `result_shape[n]`
    is the max result set size across all `0...n-1` dimensions.
)doc");

REGISTER_OP("DenseToSparseSetOperation")
    .Input("set1: T")
    .Input("set2_indices: int64")
    .Input("set2_values: T")
    .Input("set2_shape: int64")
    .Attr("set_operation: string")
    .Attr("validate_indices: bool = true")
    .Attr("T: {int8, int16, int32, int64, uint8, uint16, string}")
    .Output("result_indices: int64")
    .Output("result_values: T")
    .Output("result_shape: int64")
    .SetShapeFn([](InferenceContext* c) {
      if (c->num_inputs() != 4) {
        return errors::InvalidArgument("len(inputs) != 4.");
      }
      // The following should stay in sync with `ComputeDenseToSparse` shape
      // assertions in kernels/set_kernels.cc.
      // Dimension n contains the set values to be compared, so ranks and the
      // first n-1 dimensions of inputs and output must match.
      DimensionHandle output_rank;
      ShapeHandle input0_shape = c->input(0);
      if (c->RankKnown(input0_shape)) {
        const int32 input0_rank = c->Rank(input0_shape);
        if (input0_rank < 2) {
          return errors::InvalidArgument("Input 0, expected rank >= 2, got ",
                                         input0_rank, ".");
        }
        output_rank = c->MakeDim(input0_rank);
      } else {
        output_rank = c->UnknownDim();
      }
      TF_RETURN_IF_ERROR(
          c->ValidateSparseTensor(c->input(1), c->input(2), c->input(3)));
      DimensionHandle output_num_elements = c->Dim(input0_shape, 0);
      if (!c->ValueKnown(output_num_elements)) {
        output_num_elements = c->UnknownDim();
      }

      c->set_output(0, c->Matrix(output_num_elements, output_rank));
      c->set_output(1, c->Vector(output_num_elements));
      c->set_output(2, c->Vector(output_rank));
      return Status::OK();
    })
    .Doc(R"doc(
Applies set operation along last dimension of `Tensor` and `SparseTensor`.

See SetOperationOp::SetOperationFromContext for values of `set_operation`.

Input `set2` is a `SparseTensor` represented by `set2_indices`, `set2_values`,
and `set2_shape`. For `set2` ranked `n`, 1st `n-1` dimensions must be the same
as `set1`. Dimension `n` contains values in a set, duplicates are allowed but
ignored.

If `validate_indices` is `True`, this op validates the order and range of `set2`
indices.

Output `result` is a `SparseTensor` represented by `result_indices`,
`result_values`, and `result_shape`. For `set1` and `set2` ranked `n`, this
has rank `n` and the same 1st `n-1` dimensions as `set1` and `set2`. The `nth`
dimension contains the result of `set_operation` applied to the corresponding
`[0...n-1]` dimension of `set`.

set1: `Tensor` with rank `n`. 1st `n-1` dimensions must be the same as `set2`.
    Dimension `n` contains values in a set, duplicates are allowed but ignored.
set2_indices: 2D `Tensor`, indices of a `SparseTensor`. Must be in row-major
    order.
set2_values: 1D `Tensor`, values of a `SparseTensor`. Must be in row-major
    order.
set2_shape: 1D `Tensor`, shape of a `SparseTensor`. `set2_shape[0...n-1]` must
    be the same as the 1st `n-1` dimensions of `set1`, `result_shape[n]` is the
    max set size across `n-1` dimensions.
result_indices: 2D indices of a `SparseTensor`.
result_values: 1D values of a `SparseTensor`.
result_shape: 1D `Tensor` shape of a `SparseTensor`. `result_shape[0...n-1]` is
    the same as the 1st `n-1` dimensions of `set1` and `set2`, `result_shape[n]`
    is the max result set size across all `0...n-1` dimensions.
)doc");

REGISTER_OP("SparseToSparseSetOperation")
    .Input("set1_indices: int64")
    .Input("set1_values: T")
    .Input("set1_shape: int64")
    .Input("set2_indices: int64")
    .Input("set2_values: T")
    .Input("set2_shape: int64")
    .Attr("set_operation: string")
    .Attr("validate_indices: bool = true")
    .Attr("T: {int8, int16, int32, int64, uint8, uint16, string}")
    .Output("result_indices: int64")
    .Output("result_values: T")
    .Output("result_shape: int64")
    .SetShapeFn([](InferenceContext* c) {
      if (c->num_inputs() != 6) {
        return errors::InvalidArgument("len(inputs) != 6.");
      }
      // The following should stay in sync with `ComputeSparseToSparse` shape
      // assertions in kernels/set_kernels.cc.
      TF_RETURN_IF_ERROR(
          c->ValidateSparseTensor(c->input(0), c->input(1), c->input(2)));
      TF_RETURN_IF_ERROR(
          c->ValidateSparseTensor(c->input(3), c->input(4), c->input(5)));
      c->set_output(0, c->Matrix(c->UnknownDim(), c->UnknownDim()));
      c->set_output(1, c->Vector(c->UnknownDim()));
      c->set_output(2, c->Vector(c->UnknownDim()));
      return Status::OK();
    })
    .Doc(R"doc(
Applies set operation along last dimension of 2 `SparseTensor` inputs.

See SetOperationOp::SetOperationFromContext for values of `set_operation`.

If `validate_indices` is `True`, `SparseToSparseSetOperation` validates the
order and range of `set1` and `set2` indices.

Input `set1` is a `SparseTensor` represented by `set1_indices`, `set1_values`,
and `set1_shape`. For `set1` ranked `n`, 1st `n-1` dimensions must be the same
as `set2`. Dimension `n` contains values in a set, duplicates are allowed but
ignored.

Input `set2` is a `SparseTensor` represented by `set2_indices`, `set2_values`,
and `set2_shape`. For `set2` ranked `n`, 1st `n-1` dimensions must be the same
as `set1`. Dimension `n` contains values in a set, duplicates are allowed but
ignored.

If `validate_indices` is `True`, this op validates the order and range of `set1`
and `set2` indices.

Output `result` is a `SparseTensor` represented by `result_indices`,
`result_values`, and `result_shape`. For `set1` and `set2` ranked `n`, this
has rank `n` and the same 1st `n-1` dimensions as `set1` and `set2`. The `nth`
dimension contains the result of `set_operation` applied to the corresponding
`[0...n-1]` dimension of `set`.

set1_indices: 2D `Tensor`, indices of a `SparseTensor`. Must be in row-major
    order.
set1_values: 1D `Tensor`, values of a `SparseTensor`. Must be in row-major
    order.
set1_shape: 1D `Tensor`, shape of a `SparseTensor`. `set1_shape[0...n-1]` must
    be the same as `set2_shape[0...n-1]`, `set1_shape[n]` is the
    max set size across `0...n-1` dimensions.
set2_indices: 2D `Tensor`, indices of a `SparseTensor`. Must be in row-major
    order.
set2_values: 1D `Tensor`, values of a `SparseTensor`. Must be in row-major
    order.
set2_shape: 1D `Tensor`, shape of a `SparseTensor`. `set2_shape[0...n-1]` must
    be the same as `set1_shape[0...n-1]`, `set2_shape[n]` is the
    max set size across `0...n-1` dimensions.
result_indices: 2D indices of a `SparseTensor`.
result_values: 1D values of a `SparseTensor`.
result_shape: 1D `Tensor` shape of a `SparseTensor`. `result_shape[0...n-1]` is
    the same as the 1st `n-1` dimensions of `set1` and `set2`, `result_shape[n]`
    is the max result set size across all `0...n-1` dimensions.
)doc");

}  // namespace tensorflow
